### ISSUE TITLE
--throttle-upload option not working with Amazon Cloud Drive

### ISSUE THREAD

**Issue by automorphism88** [duplicati/duplicati#1924]
I have:
- [x] searched open and closed issues for duplicates

It looks like this was a problem in the past, but was fixed in early 2015. I've only just started using duplicati and have only tried using it with Amazon Cloud Drive, so I'm not sure if the problem I'm experiencing is at all related to the past bug, or whether it's unique to the ACD backend or affects duplicati generally.

---
### Version info

**Duplicati Version:** 2.0.1.20_canary_2016-09-01
**Operating System:** Linux Mint 18 sarah
**Backend:** Amazon Cloud Drive
### Bug description

The --throttle-upload command line option appears to have no effect, although I have confirmed that it is being set correctly by observing the "input options" output from duplicati-cli. When starting a backup with duplicati-cli, the output correctly displays the value that was entered for --throttle-upload, but it has no effect on the actual upload speed as observed with nethogs, or by observing from the timestamps in the log how long it's taking to upload each 50 MB dblock file.

I've enabled --debug-output but haven't gotten any actual error messages in the log to report, at least not related to the bandwidth throttling. The failure of --throttle-upload to work can only be detected by using other software (in my case, nethogs) to monitor how much upload bandwidth the Duplicati.CommandLine process is actually using, or by calculating the upload speed from the log timestamps. I have tried several different values for --throttle-upload (100, 200, 300, and 400, tried both with and without "kb" after the number) and no matter what, duplicati maxes out my available upload bandwidth (4Mbps, with duplicati observed to be uploading at roughly 500KB/s via both nethogs and log timestamps, no matter what I do with --throttle-upload).

I do see the put operations fail relatively often, generally either a GATEWAY_TIMEOUT error or sometimes "Error writing request" or "The socket has been shut down" but I think these errors are linked to the bandwidth throttling issue, since they seem to fit with the general flakiness of my connection when an application is trying to use as much upload bandwidth as it can. The same thing happens to my connection when I allow other applications to use as much upload bandwidth as they can, which is why I generally throttle them so that they can't use more than 80-90% of what's available.

I've also noticed that after a file upload is complete, there is sometimes a substantial delay between when the command line output (and the log, observed with tail -f) say it is starting the next upload and when I actually see it start to use the bandwidth in nethogs. Usually the delay only lasts a second or two, but I've seen it sit there for over a minute using zero bandwidth, before the next file actually starts. 

The one pattern I have noticed with these put operation errors is that they seem to occur more often when I've just started duplicati, or when I'm trying to do something else that uses my internet connection at the same time, whereas if I leave it running overnight with nothing else happening, the errors become much less frequent, and it might successfully upload a dozen or two files without getting any errors (or, judging by the timestamps, any delays of more than a couple of seconds in between files). This seems consistent with the hypothesis that the lack of upload throttling is responsible for the put errors, by causing the connection generally to become flaky, as does the fact that uploads to ACD go much more smoothly using rclone or acdcli with upload throttling enabled - it isn't perfect and I might have an occasional upload fail, but nowhere near the frequency I'm seeing with duplicati.
### Steps to reproduce
- use --throttle-upload to set a limit on upload bandwidth with duplicati-cli uploading to Amazon Cloud Drive
- use nethogs or other network monitoring software to observe the actual bandwidth used by the "Duplicati.CommandLine" process while uploading a file
- observe from timestamps in log that files are being uploaded faster than the throttle-upload value should allow
- observe that further reducing the throttle-upload value makes no difference to either of these measurements of actual bandwidth being used

**Actual result:** Duplicati uses all available bandwidth to upload files to Amazon Cloud Drive, regardless of the --throttle-upload parameter
**Expected result:** Bandwidth usage should stay reasonably close to the maximum set with --throttle-upload, like it does for me when I use rclone to upload to ACD
### debug log

Sample output verifying that throttle-upload was set (but not obeyed):

```
Input command: backup
Input arguments: 
        amzcd://[removed]
        /storage/btrfs

Input options: 
backup-name: 1
dbpath: /storage/shared/duplicati/1.sqlite
encryption-module: gpg
compression-module: zip
dblock-size: 50mb
passphrase: [removed]
zip-compression-level: 9
log-level: Information
log-file: /storage/shared/duplicati/1.log
log-retention: 1M
disable-module: console-password-input
verbose: true
asynchronous-upload-limit: 16
asynchronous-upload-folder: /storage/tmp/duplicati
tempdir: /storage/tmp/duplicati
upload-verification-file: true
debug-output: true
debug-retry-errors: true
throttle-upload: 200kb

```

Sample put error from log:

```
2016-09-06 13:16:08Z - Information: Backend event: Put - Started: duplicati-b2cdc0cb6e56549cbb22ad110569b58b4.dblock.zip.gpg (50.07 MB)
2016-09-06 13:19:03Z - Warning: Operation Put with file duplicati-b2cdc0cb6e56549cbb22ad110569b58b4.dblock.zip.gpg attempt 1 of 5 failed with message: GATEWAY_TIMEOUT
System.Net.WebException: GATEWAY_TIMEOUT
  at Duplicati.Library.Utility.AsyncHttpRequest+AsyncWrapper.GetResponseOrStream () <0x401d6840 + 0x0010b> in <filename unknown>:0 
  at Duplicati.Library.Utility.AsyncHttpRequest.GetResponse () <0x4028cd50 + 0x00163> in <filename unknown>:0 
  at Duplicati.Library.JSONWebHelper.GetResponse (Duplicati.Library.Utility.AsyncHttpRequest req, System.Object requestdata) <0x4028c570 + 0x003fb> in <filename unknown>:0 

2016-09-06 13:19:03Z - Information: Backend event: Put - Retrying: duplicati-b2cdc0cb6e56549cbb22ad110569b58b4.dblock.zip.gpg (50.07 MB)
2016-09-06 13:19:03Z - Information: Backend event: Put - Rename: duplicati-b2cdc0cb6e56549cbb22ad110569b58b4.dblock.zip.gpg (50.07 MB)
2016-09-06 13:19:03Z - Information: Backend event: Put - Rename: duplicati-b3973dbc5cf3f4e30be37f675ad963577.dblock.zip.gpg (50.07 MB)
2016-09-06 13:19:03Z - Information: Renaming "duplicati-b2cdc0cb6e56549cbb22ad110569b58b4.dblock.zip.gpg" to "duplicati-b3973dbc5cf3f4e30be37f675ad963577.dblock.zip.gpg"
2016-09-06 13:19:13Z - Information: Backend event: Put - Started: duplicati-b3973dbc5cf3f4e30be37f675ad963577.dblock.zip.gpg (50.07 MB)
2016-09-06 13:21:13Z - Information: Backend event: Put - Completed: duplicati-b3973dbc5cf3f4e30be37f675ad963577.dblock.zip.gpg (50.07 MB)

```

You can see from the log that the file uploaded successfully on the second try after a GATEWAY_TIMEOUT on the first try, and took exactly two minutes, which reflects a pause of about twenty seconds between when the log says the file started uploading and when it actually started. When things are running more smoothly, the delay with zero bandwidth being used between when duplicity reports that it's starting the upload and when it actually starts will only last a couple of seconds, resulting in a log that looks more like the following, where a dblock file of the same size took 1:42 to finish:

```
2016-09-06 13:00:35Z - Information: Backend event: Put - Started: duplicati-befb1323f91dd45849acc0bf375cc5d91.dblock.zip.gpg (50.07 MB)
2016-09-06 13:02:17Z - Information: Backend event: Put - Completed: duplicati-befb1323f91dd45849acc0bf375cc5d91.dblock.zip.gpg (50.07 MB)
```

From watching nethogs, I know this difference is due to the upload taking longer to start, and not to any difference in speed once it does. Also, during both of these uploads, throttle-upload was set to "200kb" as shown above (though I see the exact same results if it's set to something smaller or larger, or not set at all), while the timestamps above are evidence that the files are actually being uploaded at 500 KB/s, as I would expect them to if duplicati is maxing out my connection's 4Mbps upload capacity. At 500 KB/s, it should take 100 seconds, or 1:40, to upload a 50MB file, plus the second or two delay observed with nethogs with bandwidth use at zero before the upload actually starts, which is exactly what these timestamps show.
##

**Comment by tophee** [comment]
While I have not done any of the kind of testing done by automorphism88, I can confirm that the throttling doesn't work with ACD, simply by looking at the Task Manager in Windows which tells me that duplicati is using the maximum available bandwidth, no matter what I enter as a limit.

**Comment by renestach** [comment]
We know that the throttle is broken at the moment. https://github.com/duplicati/duplicati/issues/1729

**Comment by dominicsayers** [comment]
:+1:

As a new user I totally love Duplicati 2.0.

Throttling would make it pretty much perfect. I hope you can find some time to fix this.

**Comment by jerrac** [comment]
I just ran into this as well. Uploading to Amazon Cloud pegs my upload bandwidth.

I actually ran into this issue with the actual Amazon Cloud windows app years ago. They've since fixed it, but at the time I learned that a flooded upload bandwidth makes it impossible to actually surf the net at the same time. No way to send tcp packets. Thus, eventually, any upload to AC would stop working as well. I assume that's the "Gateway 504" errors I've seen from Duplicati.

So, a working throttle would be really really nice. :)

**Comment by jerrac** [comment]
As we wait for this issue to get fixed, has anyone figured out a work around?

I have an OPNSense firewall. I'm currently poking around to see if there's a way to throttle uploads to AC using it. 

Anyone know of any other tools that could have the same effect?

**Comment by kenkendk** [comment]
I just verified that throttling is working as expected in my tests when writing to a file destination.

Unfortunately, I cannot test with ACD anymore as they have changed the terms, so I now need a German bank account to use the "free trial" mode (same for the paid version).

The updates to the throttle code were implemented in 0190a9994e55c0e4c6b7c7eaac43ec25dd811fc5 (Oct. 17th 2016).

@jerrac: You appear to have the issue as well, and I assume you are running a recent version. 
Are you using the GUI or the commandline? If you are using the GUI, have you set the throttle options on the backup itself (i.e. not in the general options under "Settings")?

**Comment by jerrac** [comment]
I just checked the version. I was on 2.0.1.45_experimental_2017-02-03. 

After upgrading to 2.0.1.49_experimental_2017-02-23 I ran the job again. It's not obeying my throttle option.

I'm using the Windows 10 GUI. I've applied the setting to the job itself, not globally. 

I'll try globally later, but I'd rather not keep it that way. Some of my jobs are on my local network only, not going out to the web. So there's no reason to throttle them.

**Comment by kenkendk** [comment]
There is no need to try with the global setting, the local one should work.

**Comment by kenkendk** [comment]
Can you try with a local folder as the destination and a throttle applied, so we can reduce the problem to "something in ACD" ?

**Comment by jerrac** [comment]
This is what I've been looking at when checking bandwidth usage:
![taskmanagerscreen](https://cloud.githubusercontent.com/assets/910764/23516580/77b8c10e-ff22-11e6-94f5-9d2120645142.gif)

Is there a better way to test this than just watching Task Manager? I'd really rather have a log file or something to upload...

I've also checked the process tab. The Duplicati.GUI.TrayIcon process shows more than the throttled bandwidth.

I did try the global setting, there was no effect. 

The screenshot is from my attempt at throttling an ftp upload. That particular job goes to a Seagate GoFlex Home NAS sitting on my LAN. I set ```--throttle-upload=1MB```.

So, it appears this is not an Amazon Cloud Drive Specific issue. 

> Can you try with a local folder as the destination and a throttle applied, so we can reduce the problem to "something in ACD" ?

Just in case my FTP job wasn't what you were after, did you mean another folder on my local hard disk? I wouldn't think the throttle would effect that kind of destination since it is not over a network. And I'm not sure how I would monitor that.

**Comment by kenkendk** [comment]
If you view the live log in Duplicati, and set the level to "Profiling", you should see messages such as "Uploaded xxmb in yy (zz mb/s)".

I did mean a local folder, but I can test with FTP without too much trouble.
I will see if I can reproduce the unlimited transfer speed to an FTP server.

**Comment by kenkendk** [comment]
I just tested this today with an FTP server and `--upload-limit=500kb`.
The upload was throttled correctly, giving just under 500kb/s transfer speeds.

This gave results such as:
```
Uploaded 61,47 MB in 00:02:07.1718490, 494,96 KB/s
```

I used the `aftp` backend instead of the normal `ftp` as it supports some extra SSL options that I needed.

**Comment by kenkendk** [comment]
If I export the backup "As commandline", I see ` --throttle-upload="500KB"`, can you check what you get?

**Comment by jerrac** [comment]
For this configuration:

    "C:\Program Files\Duplicati 2\Duplicati.CommandLine.exe" backup "ftp://path/to/backups?auth-username=username&auth-password=password" "T:\WhatImBackingUp\\" --backup-name="Name" --dbpath="C:\Users\USER\AppData\Roaming\Duplicati\ACEYPXZAHE.sqlite" --encryption-module="aes" --compression-module="zip" --dblock-size="50mb" --passphrase="pass" --throttle-upload="500KB" --disable-module="console-password-input"

I see one log messages like this:

    Mar 7, 2017 10:44 AM: Uploaded 398.09 KB in 00:00:00.2212136, 1.76 MB/s

Most are like this, where the speed is close to the 500KB limit:

    Mar 7, 2017 10:44 AM: Uploaded 1.32 MB in 00:00:02.2602505, 597.28 KB/s

But, while watching Task Manager, ```Duplicati.GUI.TrayIcon``` was consistently using 4.0 mbps in the network column.

So, is there a reason that Task Manager would report a different speed than Duplicati?

**Comment by kenkendk** [comment]
The "faster" one is just 400kb, so the throttle does not have time to kick in.

Since Duplicati is just uploading one file at a time, I do not have a good explanation for 4MB/s.
On my MacBook, the network usage was reported around 500kb/s during the upload in the Activity Monitor tool (similar to Task Manager).

**Comment by jerrac** [comment]
Interesting.

I'm currently running a backup with the same configuration, just sending
it to Amazon Cloud instead of my ftp... I'm seeing similar behavior in
the logs. Duplicati seems to say it's obeying the limit, but Task
Manager shows the icon process using 3+ Mbps of network.

I'll see about finding time to run these tests on Ubuntu. We'll see what
System Monitor says, will it mirror Task Manager or Activity Monitor? :)

**Comment by renestach** [comment]
Just to make sure. 400kBytes are about 3.2 Mbps. You did not mix up the units, did you?

**Comment by jerrac** [comment]
Oh wow... That explains a lot... I completely forgot to check the units. And I did mix them up. 

When I was first having problems with my network getting pegged by uploads, I was using the MBytes/s option. My upload limit is around 10Mbps, so I set it to 5 MBytes/s. Which I know see is way higher than my actual limit... And that's why I thought the throttle was not working. 

Man, that's a rather embarrassing oversight. I know about bits vs. bytes... Heh.

Apologies for putting you through the hassle of troubleshooting this...
